### PR TITLE
Added Swashbuckle.AspNetCore.Annotations.SwaggerIgnoreAttribute

### PR DESCRIPTION
**Pull request by jcracknell**
Application of the SwaggerIgnoreAttribute suppresses the annotated property during schema generation, without altering the serialization behavior as with JsonIgnoreAttribute.

This is useful in a scenario where you want to capture and reject unhandled properties, without advertising that you accept `additionalProperties` (which is the opposite of what is intended):

```cs
public abstract class RequestBase : IValidatableObject {
    [SwaggerIgnore]
    [JsonExtensionData]
    public IDictionary<string, JsonElement>? UnsupportedProperties { get; set; }

    public virtual IEnumerable<ValidationResult> Validate(ValidationContext validationContext) {
        if(UnsupportedProperties is { Count: not 0 })
            foreach(var entry in UnsupportedProperties)
                yield return new ValidationResult($"Unsupported property: {entry.Key}.", new[] { entry.Key });
    }
}
```

I added the annotation to the Swashbuckle.AspNetCore.SwaggerGen project as it is much easier to filter the properties in the JsonSerializerDataContractResolver than after the fact in a filter, as you would need to regenerate the data contract for the type and then remove all of the annotated properties and derived side effects.

### CODE REVIEW

**Comment by Havunen** [comment]
Maybe the attribute should be called something else from "ignore" ie. SwaggerHiddenAttribute to indicate it is not the same thing.

**Comment by martincostello** [comment]
Thanks for contributing - if you'd like to continue with this pull request, please rebase against the default branch to pick up our new CI.

**Comment by jcracknell** [comment]
Done. I've also added commits to support the attribute on operations and parameters.

I'm aware of the current recommended approach of excluding methods using ApiExplorerSettingsAttribute, but there's something to be said for consistency, and there's also a difference between excluding the operation from Swashbuckle and excluding it from the ApiExplorer.

File: src/Swashbuckle.AspNetCore.SwaggerGen/Annotations/SwaggerIgnoreAttribute.cs
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerGen/Annotations/SwaggerIgnoreAttribute.cs:14]: System.Text.Json also supports fields for serialisation, so should this too?

File: src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGenerator.cs
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGenerator.cs:87]: Could this be `!HasAttribute<T>()`? I just saw it in the previous file and it looks terser than this.

File: test/Swashbuckle.AspNetCore.SwaggerGen.Test/Fixtures/SwaggerIngoreAnnotatedType.cs
Comment (martincostello) [inline test/Swashbuckle.AspNetCore.SwaggerGen.Test/Fixtures/SwaggerIngoreAnnotatedType.cs:6]:
```suggestion
namespace Swashbuckle.AspNetCore.SwaggerGen.Test.Fixtures
{
```

File: test/Swashbuckle.AspNetCore.SwaggerGen.Test/SchemaGenerator/JsonSerializerSchemaGeneratorTests.cs
Comment (martincostello) [inline test/Swashbuckle.AspNetCore.SwaggerGen.Test/SchemaGenerator/JsonSerializerSchemaGeneratorTests.cs:791]:
```suggestion
        public void GenerateSchema_HonorsAttribute_SwaggerIgnore()
        {
```

File: src/Swashbuckle.AspNetCore.SwaggerGen/Annotations/SwaggerIgnoreAttribute.cs
Comment (jcracknell) [reply src/Swashbuckle.AspNetCore.SwaggerGen/Annotations/SwaggerIgnoreAttribute.cs:14]: I actually didn't realize JsonSerializerOptions had been added for this in .NET 5, however it looks as though significant portions of the Swashbuckle codebase only consider properties (e.g. looking at the CustomAttributes extension used elsewhere), so I'm hesitant to take on this change.

File: src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGenerator.cs
Comment (jcracknell) [reply src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGenerator.cs]: The behavior of this particular extension appears to be provided by the builtin PropertyInfo.IsDefined(Type). I've removed this extension on PropertyInfo, and added extensions with the same name on ApiDescription and ApiParameterDescription, as in this case the extension is required to handle e.g. property and parameter sourced API parameters.

It also occurred to me that the ApiExplorer provided models include the instantiated attributes as metadata, however I am unclear how this might affect synthetically generated instances in unit tests.

File: src/Swashbuckle.AspNetCore.SwaggerGen/SchemaGenerator/PropertyInfoExtensions.cs
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerGen/SchemaGenerator/PropertyInfoExtensions.cs:13]: Removing this is a breaking change because it's `public`.

File: src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/ApiDescriptionExtensions.cs
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/ApiDescriptionExtensions.cs:49]:
```suggestion
        public static bool HasAttribute(this ApiDescription apiDescription, Type attributeType)
        {
            if (!apiDescription.TryGetMethodInfo(out var methodInfo))
                return false;

            return methodInfo.IsDefined(attributeType)
                || methodInfo.DeclaringType?.IsDefined(attributeType) is true;
        }

```

File: src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/ApiDescriptionExtensions.cs
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/ApiDescriptionExtensions.cs:40]: Missing null check on `apiDescription`.

File: src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/ApiParameterDescriptionExtensions.cs
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/ApiParameterDescriptionExtensions.cs:78]:
```suggestion
            if (apiParameter.PropertyInfo() is { } propertyInfo)
                return propertyInfo.IsDefined(attributeType);
            if (apiParameter.ParameterInfo() is { } parameterInfo)
                return parameterInfo.IsDefined(attributeType);
```

File: src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/ApiParameterDescriptionExtensions.cs
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/ApiParameterDescriptionExtensions.cs:73]: Missing null check.

File: src/Swashbuckle.AspNetCore.SwaggerGen/Annotations/SwaggerIgnoreAttribute.cs
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerGen/Annotations/SwaggerIgnoreAttribute.cs:15]: We can always unseal it if a use case for extension ever appears.

```suggestion
    public sealed class SwaggerIgnoreAttribute : Attribute { }
```

**Comment by jcracknell** [comment]
You're correct that PropertyInfo.HasAttribute is public API and dropping it would be a breaking change, so I've dropped the commit. Unfortunately it's also going to be highly infectious thanks to Intellisense, so I'd suggest marking it as obsolete so people are aware, but I'll leave that up to you.

I've also dropped the commit adding extensions on Api{,Parameter}Description, as I had not realized that these are ApiExplorer types, and public extension methods are inappropriate here for the same reason.

**Review by martincostello — APPROVED** [review]

**Comment by martincostello** [comment]
Thanks for the contribution - I'm not going to merge this immediately as I'm not sure if our next release is going to be a patch release of just small changes/tweaks/fixes, or a minor release with some new features too (like this).

Once we know the next release is going to be minor (or major), I'll merge this.

**Comment by joesdu** [comment]
it's look like [this](https://github.com/joesdu/EasilyNET/blob/main/src/EasilyNET.WebCore.Swagger/SwaggerFilters/SwaggerPropertyIgnoreFilter.cs) ?

**Pull request merged**
